From f36b00205fd5f0dd8c12f5f1a104b750346fd2fe Mon Sep 17 00:00:00 2001
From: Li Lingfeng <lilingfeng3@huawei.com>
Date: Wed, 7 Dec 2022 18:39:37 +0800
Subject: [PATCH] block: check flags of claimed slave bdev to fix uaf for
 bd_holder_dir
Patch-mainline: Not yet, from openEuler
References: bsn#22
openEuler-commit: f36b00205fd5f0dd8c12f5f1a104b750346fd2fe
Modified-by-SEL: No


hulk inclusion
category: bugfix
bugzilla: https://gitee.com/openeuler/kernel/issues/I60QE9
CVE: NA

--------------------------------

As explained in 32c39e8a7613 ("block: fix use after free for
bd_holder_dir"), we should make sure the "disk" is still live and
then grab a reference to 'bd_holder_dir'. However, the "disk"
should be "the claimed slave bdev" rather than "the holding disk".

Fixes: 32c39e8a7613 ("block: fix use after free for bd_holder_dir")
Signed-off-by: Li Lingfeng <lilingfeng3@huawei.com>
Reviewed-by: Yu Kuai <yukuai3@huawei.com>
Reviewed-by: Jason Yan <yanaijie@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 fs/block_dev.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/fs/block_dev.c b/fs/block_dev.c
index 07cbe6190463..ef20ee346ec7 100644
--- a/fs/block_dev.c
+++ b/fs/block_dev.c
@@ -1271,7 +1271,7 @@ int bd_link_disk_holder(struct block_device *bdev, struct gendisk *disk)
 	 * the holder directory.  Hold on to it.
 	 */
 	down_read(&bdev->bd_disk->lookup_sem);
-	if (!(disk->flags & GENHD_FL_UP)) {
+	if (!(bdev->bd_disk->flags & GENHD_FL_UP)) {
 		up_read(&bdev->bd_disk->lookup_sem);
 		return -ENODEV;
 	}
-- 
2.33.0

